### PR TITLE
Run validation on tests and fix hydration output validation

### PR DESCRIPTION
**Pull request by gnawf**

### CODE REVIEW

File: .github/workflows/pull_request.yml [hunk 1]
```diff
@@ -18,3 +18,10 @@ jobs:
           java-version: '11.0.9'
       - name: build and test
         run: ./gradlew assemble --info && ./gradlew check --info
+      - uses: actions/upload-artifact@v3
```
Comment (gnawf) [inline .github/workflows/pull_request.yml:21]: Uploads the test report if it failed

File: engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelHydrationValidation.kt [hunk 2]
```diff
@@ -91,40 +74,44 @@ internal class NadelHydrationValidation(
...
+            is GraphQLUnionType -> actorOutputType.types
+            is GraphQLInterfaceType -> overallSchema.getImplementations(actorOutputType)
+            else -> listOf(actorOutputType as GraphQLNamedOutputType)
+        }
+
+        val typeValidation = actorOutputTypes
+            // Find incompatible output types
+            .filter { actorOutputType ->
+                acceptableOutputTypes.none { acceptableOutputType ->
+                    typeValidation.isAssignableTo(lhs = acceptableOutputType, rhs = actorOutputType)
```
Comment (gnawf) [inline engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelHydrationValidation.kt:104]: Made some changes to the way the output type is validated. Basically we can validate stuff like

```graphql
type Query {
  userById(id: ID!): User
}
type Issue {
  assignee: AbstractUser @hydrated(field: "userById")
}
interface AbstractUser {
}
type User implements AbstractUser {
}
```

i.e. type doesn't exactly match up, but we do a smarter check on whether they're compatible.

File: engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelHydrationValidation.kt [hunk 3]
```diff
@@ -58,29 +56,14 @@ internal class NadelHydrationValidation(
...
-            val actorServiceQueryType = actorService.underlyingSchema.queryType
-            val actorField = actorServiceQueryType.getFieldAt(hydration.pathToActorField)
+            val actorField = overallSchema.queryType.getFieldAt(hydration.pathToActorField)
             if (actorField == null) {
-                errors.add(MissingHydrationActorField(parent, overallField, hydration, actorServiceQueryType))
+                errors.add(MissingHydrationActorField(parent, overallField, hydration))
                 continue
             }
 
-            val overallActorField = overallSchema.queryType.getFieldAt(hydration.pathToActorField)
```
Comment (gnawf) [inline engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelHydrationValidation.kt:68]: Removed this as we simply grab from the overall schema now.

File: engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelHydrationValidation.kt [hunk 4]
```diff
@@ -91,40 +74,44 @@ internal class NadelHydrationValidation(
...
-                            parent,
-                            overallField
-                        )
-                    )
-                overallType = overallSchema.getType(overallTypeName) as GraphQLUnmodifiedType
-            } else {
-                return listOf(FieldWithPolymorphicHydrationMustReturnAUnion(parent, overallField))
             }
-        }
-        val typeValidation = typeValidation.validate(
```
Comment (gnawf) [inline engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelHydrationValidation.kt:121]: No need to run type validation here as we no longer deal with two type where one is underlying and one is overall.

The validation for the types is already done elsewhere. So in the code here, if they're assignable to each other, then the types are valid.

File: engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelSchemaValidationError.kt [hunk 5]
```diff
@@ -59,6 +61,21 @@ sealed interface NadelSchemaValidationError {
             .build()
     }
 
+    data class UnionHasExtraType(
```
Comment (gnawf) [inline engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelSchemaValidationError.kt:64]: Not actually used right now.

File: engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelSchemaValidationError.kt [hunk 6]
```diff
@@ -226,32 +243,14 @@ sealed interface NadelSchemaValidationError {
...
             val of = makeFieldCoordinates(parentType.overall.name, overallField.name)
             val s = hydration.serviceName
-            val af = "${actorServiceQueryType.name}.${hydration.pathToActorField.joinToString(separator = ".")}"
-            "Field $of tried to hydrate from non-existent field $af on service $s"
-        }
-
-        override val subject = overallField
-    }
-
-    data class MissingHydrationActorFieldInOverallSchema(
```
Comment (gnawf) [inline engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelSchemaValidationError.kt:243]: I don't think we actually reference these explicitly, so I don't think removing the type matters.

File: engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelTypeValidation.kt [hunk 7]
```diff
@@ -98,6 +106,31 @@ internal class NadelTypeValidation(
...
+     * ```
+     *
+     * Note: this assumes both types are from the same schema. This does NOT
+     * deal with differences between overall and underlying schema.
+     */
+    fun isAssignableTo(lhs: GraphQLNamedOutputType, rhs: GraphQLNamedOutputType): Boolean {
+        if (lhs.name == rhs.name) {
+            return true
+        }
+        if (lhs.name == GraphQLID.name && rhs.name == GraphQLString.name) {
```
Comment (gnawf) [inline engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelTypeValidation.kt:125]: I'm going to leave this in… I think the other type validation should follow this in the future? Thoughts?

cc: @felipe-gdr

File: engine-nextgen/src/test/kotlin/graphql/nadel/validation/NadelHydrationValidationTest.kt [hunk 8]
```diff
@@ -914,11 +914,295 @@ class NadelHydrationValidationTest : DescribeSpec({
...
+            val errors = validate(fixture)
+            assert(errors.isNotEmpty())
+            val error = errors.singleOfType<HydrationIncompatibleOutputType>()
+            assert(error.parentType.overall.name == "Issue")
+            assert(error.actorField.name == "externalUser")
+            assert((error.actorField.type as GraphQLNamedType).name == "ExternalUser")
+            assert(error.incompatibleOutputType.name == "ExternalUser")
+        }
+
+        it("fails if actor output type does not implement interface") {
```
Comment (gnawf) [inline engine-nextgen/src/test/kotlin/graphql/nadel/validation/NadelHydrationValidationTest.kt:995]: Some tests against the is hydration type assignable code I added.

File: test/src/test/kotlin/graphql/nadel/tests/EngineTestHook.kt [hunk 9]
```diff
@@ -43,6 +44,10 @@ interface EngineTestHook {
         return builder
     }
 
+    fun isSchemaValid(errors: Set<NadelSchemaValidationError>): Boolean {
```
Comment (gnawf) [inline test/src/test/kotlin/graphql/nadel/tests/EngineTestHook.kt:47]: Hook to bypass validation e.g. to ensure we let the malformed JSW schema through.

See the test `let-jsw-do-jsw-things.yml`

File: test/src/test/resources/fixtures/basic/schema-transformation-is-applied.yml [hunk 10]
```diff
@@ -22,13 +22,18 @@ underlyingSchema:
       hello: String
     }
 
+    type Subscription {
```
Comment (gnawf) [inline test/src/test/resources/fixtures/basic/schema-transformation-is-applied.yml:25]: All changes beyond here are just fixes to the fixtures to make them valid.

File: engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelUnionValidation.kt [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+internal class NadelUnionValidation(
+    private val typeValidation: NadelTypeValidation,
+) {
+    fun validate(
+        schemaElement: NadelServiceSchemaElement,
+    ): List<NadelSchemaValidationError> {
+        return emptyList()
+
+        // return if (schemaElement.overall is GraphQLUnionType && schemaElement.underlying is GraphQLUnionType) {
```
Comment (gnawf) [inline engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelUnionValidation.kt:11]: Couldn't quite wrap my head around the logic here. Will revisit later.

File: test/src/test/kotlin/graphql/nadel/tests/EngineTests.kt [hunk 12]
```diff
@@ -309,3 +324,59 @@ private suspend fun execute(
...
+    val nadel = Nadel.newNadel()
+        .engineFactory {
+            object : NadelExecutionEngine {
+                override fun execute(
+                    executionInput: ExecutionInput,
+                    queryDocument: Document,
+                    instrumentationState: InstrumentationState?,
+                    nadelExecutionParams: NadelExecutionParams,
+                ): CompletableFuture<ExecutionResult> {
+                    throw UnsupportedOperationException("no-op")
```
Comment (gnawf) [inline test/src/test/kotlin/graphql/nadel/tests/EngineTests.kt:341]: I mentioned this to @felipe-gdr but in the future I think we should split the creation of the "engine schema" and "services" from the `Nadel` object e.g.

1. Introduce new `NadelSchemaBuilder`
2. Create engine schema and services using `NadelSchemaBuilder` say the output is some `NadelSchemas` object
3. Make Nadel.Builder take in `NadelSchemas` _or_ optionally work as today, but delegate off to `NadelSchemaBuilder`

The issue today is creating a Nadel object is very closely tied to creating a `NextgenEngine` which then creates a blueprint, and the blueprint assumes a valid schema. So for validation, to get the engine schema and services out of the Nadel object we need to stub the engine so no blueprint is instantiated.

Comment (gnawf) [inline test/src/test/kotlin/graphql/nadel/tests/EngineTests.kt:341]: I mentioned this to @felipe-gdr but in the future I think we should split the creation of the "engine schema" and "services" from the `Nadel` object e.g.

1. Introduce new `NadelSchemaBuilder`
2. Create engine schema and services using `NadelSchemaBuilder` say the output is some `NadelSchemas` object
3. Make Nadel.Builder take in `NadelSchemas` _or_ optionally work as today, but delegate off to `NadelSchemaBuilder`

The issue today is creating a Nadel object is very closely tied to creating a `NextgenEngine` which then creates a blueprint, and the blueprint assumes a valid schema. So to get the engine schema and services out of the Nadel object we need to stub the engine so no blueprint is instantiated.

File: engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelTypeValidation.kt (hunk 7 above)
Comment (faizan-siddiqui) [reply engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelTypeValidation.kt:125]: so this allows String -> ID, and one way only?

Comment (gnawf) [reply engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelTypeValidation.kt:125]: Yep. I think the other way around is technically possible too. They're both Strings, but as an input type ID accepts Ints too.

File: test/src/test/resources/fixtures/renames/types/renamed-type-is-shared-but-not-renamed-in-one-service.yml [hunk 13]
```diff
@@ -1,5 +1,6 @@
 name: renamed type is shared but not renamed in one service
-enabled: true
+# Not supported
```
Comment (temaEmelyan) [inline test/src/test/resources/fixtures/renames/types/renamed-type-is-shared-but-not-renamed-in-one-service.yml:2]: what's the plan with these? are we going to delete them

File: engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelUnionValidation.kt (hunk 11 above)
Comment (temaEmelyan) [reply engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelUnionValidation.kt:11]: We have this same code in the gateway, right? Are we leaving a gap in our validation by commenting this out?

File: test/src/test/resources/fixtures/renames/types/renamed-type-is-shared-but-not-renamed-in-one-service.yml (hunk 13 above)
Comment (gnawf) [reply test/src/test/resources/fixtures/renames/types/renamed-type-is-shared-but-not-renamed-in-one-service.yml:2]: I think we'll want to support it eventually so I'm keeping the test in. This is the shared rename issue where we'll want the repeatable

```
```

syntax.

File: engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelUnionValidation.kt (hunk 11 above)
Comment (gnawf) [reply engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelUnionValidation.kt:11]: I don't see any validation on it, so don't think it's a big deal.

**Review by temaEmelyan — APPROVED** [review]

**Pull request merged**